internal/overlord/servstate/handlers.go

```diff
@@ -51,183 +54,528 @@ var (
+	stateKilling
+	stateStopped
+	stateBackoffWait
 )
 
-// Start starts the named service after also starting all of its dependencies.
+func (s serviceState) String() string {
+	switch s {
+	case stateInitial:
+		return "initial"
+	case stateStarting:
+		return "starting"
+	case stateRunning:
+		return "running"
+	case stateBackoffWait:
+		return "backoff-wait"
+	case stateTerminating:
+		return "terminating"
+	case stateKilling:
+		return "killing"
+	case stateStopped:
+		return "stopped"
+	default:
+		return "unknown"
+	}
+}
+
+// service holds the information for a service under our control.
+type service struct {
+	lock         sync.Mutex
+	manager      *ServiceManager
+	state        serviceState
+	config       *plan.Service
+	logs         *servicelog.RingBuffer
+	started      chan *int
+	stopped      chan error
+	cmd          *exec.Cmd
+	backoffIndex int
+	resetTimer   *time.Timer
+}
+
 func (m *ServiceManager) doStart(task *state.Task, tomb *tomb.Tomb) error {
+	m.state.Lock()
+	request, err := TaskServiceRequest(task)
+	m.state.Unlock()
+	if err != nil {
+		return fmt.Errorf("cannot get service request: %w", err)
+	}
+
 	releasePlan, err := m.acquirePlan()
 	if err != nil {
+		return fmt.Errorf("cannot acquire plan lock: %w", err)
+	}
+	config, ok := m.plan.Services[request.Name]
+	if !ok {
+		releasePlan()
+		return fmt.Errorf("cannot find service %q in plan", request.Name)
+	}
+	releasePlan()
+
+	m.servicesLock.Lock()
+	s := m.services[config.Name]
+	if s != nil && s.getStateLocked() != stateStopped {
+		m.servicesLock.Unlock()
+		m.state.Lock()
+		task.Logf("Service %q already started.", config.Name)
+		m.state.Unlock()
+		return nil
+	}
+	if s == nil {
+		s = &service{
+			manager: m,
+			state:   stateInitial,
+			config:  config.Copy(),
+			logs:    servicelog.NewRingBuffer(maxLogBytes),
+		}
+		m.services[config.Name] = s
+	} else {
+		s.backoffIndex = 0
+		s.transition(stateInitial)
+	}
+	s.started = make(chan *int, 1)
+	m.servicesLock.Unlock()
+
+	// Start the service and transition to stateStarting.
+	err = s.start()
+	if err != nil {
+		m.removeService(config.Name)
 		return err
 	}
-	defer releasePlan()
 
+	// Wait for a small amount of time, and if the service hasn't exited,
+	// consider it a success.
+	timeout := time.After(okayWait + 100*time.Millisecond)
+	for {
+		select {
+		case exitCode := <-s.started:
+			if exitCode != nil {
+				m.removeService(config.Name)
+				return fmt.Errorf("cannot start service: exited quickly with code %d", *exitCode)
+			}
+			// Started successfully (ran for small amount of time without exiting).
+			return nil
+		case <-tomb.Dying():
+			// Start cancelled (shouldn't really happen, so disallow for simplicity).
+			logger.Noticef("Cannot cancel start (service %q)", config.Name)
+		case <-timeout:
+			// Should never happen, but don't block just in case we get something wrong.
+			m.removeService(config.Name)
+			return fmt.Errorf("internal error: timed out waiting for start")
```

Comment:
Yeah, it's because the state machine should always send to the `started` channel, either in `okayWaitElapsed` after it's been running fine for 1s, or when it exits within 1s. So this is only to catch a coding/logic bug. I've clarified in the comments (for the stop case as well).